lib/src/block_syntaxes/list_syntax.dart

```diff
@@ -89,16 +99,41 @@ abstract class ListSyntax extends BlockSyntax {
       } else if (tryMatch(hrPattern)) {
         // Horizontal rule takes precedence to a new list item.
         break;
-      } else if (tryMatch(ulPattern) || tryMatch(olPattern)) {
+      } else if ((isCheckboxListSubclass &&
+              (tryMatch(ulWithCheckBoxPattern) ||
+                  tryMatch(olWithCheckBoxPattern))) ||
+          tryMatch(ulPattern) ||
+          tryMatch(olPattern)) {
+        // The checkbox subclasse's patterns ([ulWithCheckBoxPattern] and
+        // [olWithCheckBoxPattern]) have 2 extra capturing groups at the 5
+        // position to capture the checkbox. This shift the other groups
+        // forward by 2 slots.
+        final cbGroupOffset = isCheckboxListSubclass ? 2 : 0;
```

Comment:
This is pretty clever. The code below is a little unusual (`match[6 + cbGroupOffset]`) but I think it's a good trick, and not hard to understand.